dbms/src/Storages/MergeTree/MergeTreeMinMaxIndex.cpp

```diff
@@ -0,0 +1,164 @@
+#include <Storages/MergeTree/MergeTreeMinMaxIndex.h>
+
+#include <Interpreters/ExpressionActions.h>
+#include <Interpreters/ExpressionAnalyzer.h>
+#include <Interpreters/SyntaxAnalyzer.h>
+
+#include <Poco/Logger.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+    extern const int INCORRECT_QUERY;
+}
+
+
+MergeTreeMinMaxGranule::MergeTreeMinMaxGranule(const MergeTreeMinMaxIndex & index)
+    : MergeTreeIndexGranule(), index(index), parallelogram()
+{
+}
+
+void MergeTreeMinMaxGranule::serializeBinary(WriteBuffer & ostr) const
+{
+    if (empty())
+        throw Exception(
+                "Attempt to write empty minmax index `" + index.name + "`", ErrorCodes::LOGICAL_ERROR);
+
+    for (size_t i = 0; i < index.columns.size(); ++i)
+    {
+        const DataTypePtr & type = index.data_types[i];
+
+        type->serializeBinary(parallelogram[i].left, ostr);
+        type->serializeBinary(parallelogram[i].right, ostr);
+    }
+}
+
+void MergeTreeMinMaxGranule::deserializeBinary(ReadBuffer & istr)
+{
+    parallelogram.clear();
+    for (size_t i = 0; i < index.columns.size(); ++i)
+    {
+        const DataTypePtr & type = index.data_types[i];
+
+        Field min_val;
+        type->deserializeBinary(min_val, istr);
+        Field max_val;
+        type->deserializeBinary(max_val, istr);
+
+        parallelogram.emplace_back(min_val, true, max_val, true);
+    }
+}
+
+String MergeTreeMinMaxGranule::toString() const
+{
+    String res = "";
+
+    for (size_t i = 0; i < parallelogram.size(); ++i)
+    {
+        res += "["
+                + applyVisitor(FieldVisitorToString(), parallelogram[i].left) + ", "
+                + applyVisitor(FieldVisitorToString(), parallelogram[i].right) + "]";
+    }
+
+    return res;
+}
+
+void MergeTreeMinMaxGranule::update(const Block & block, size_t * pos, size_t limit)
+{
+    size_t rows_read = std::min(limit, block.rows() - *pos);
```

Comment:
Maybe add check that `*pos` is less than number of rows? I know from outer code, that it's always true.